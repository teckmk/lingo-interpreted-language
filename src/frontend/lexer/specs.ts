import { EOL } from "os"

export enum TokenType {
  NumberLiteral = "NUMBER_LITERAL",
  StringLiteral = "STRING_LITERAL",
  Identifier = "IDENTIFIER",
  SingleLineComment = "SINLE_LINE_COMMENT",
  DocComment = "DOC_COMMENT",
  MultiLineComment = "MULTI_LINE_COMMENT",
  WhiteSpace = "WHITE_SPACE",
  Tab = "TAB",
  Indent = "INDENT",
  Dedent = "DEDENT",
  EOF = "END_OF_FILE",
  EOL = "END_OF_LINE",

  Equals = "ASSIGNMENT_OPERATOR", // =
  Comma = "COMMA", // ,
  Dot = "DOT", // .
  Exclamation = "EXLAMATION", // "!"
  Colon = "COLON", //":"
  OpenParen = "OPEN_PAREN", //  (
  CloseParen = "CLOSE_PAREN", // )
  OpenBrace = "OPEN_BRACE", // {
  CloseBrace = "CLOSE_BRACE", // }
  OpenBracket = "OPEN_BRACKET", // [
  CloseBracket = "CLOSE_BRACKET", // ]
  LogicGate = "LOGIC_GATE", // && || and or
  AdditiveOperator = "ADDITIVE_OPERATOR", //  + -
  MulitipicativeOperator = "MULTIPICATIVE_OPERATOR", // / * %
  EqualityOperator = "EQUALITY_OPERATOR", // == !=
  RelationalOperator = "RELATIONAL_OPERATOR", // > < >= <=

  // Reserved
  // KeyWords
  Let = "LET",
  Const = "CONST",
  Final = "FINAL",
  Fn = "FN",
  Return = "RETURN",
  If = "IF",
  Else = "ELSE",
  While = "WHILE",

  // Primitive Types
  NumberType = "NUMBER_TYPE",
  StringType = "STRING_TYPE",
  BooleanType = "BOOL_TYPE",
  DynamicType = "DYNAMIC_TYPE",
}

export type Spec = {
  regex: RegExp
  tokenType: TokenType
}

export const specs: Spec[] = [
  { regex: /^[+-]?([\d]*[.])?[\d]+/, tokenType: TokenType.NumberLiteral },
  { regex: /^"[^"]*"/, tokenType: TokenType.StringLiteral },
  { regex: /^'[^']*'/, tokenType: TokenType.StringLiteral },

  { regex: /^\bvar\b/, tokenType: TokenType.Let },
  { regex: /^\bconst\b/, tokenType: TokenType.Const },
  { regex: /^\bfinal\b/, tokenType: TokenType.Final },
  { regex: /^\bfn\b/, tokenType: TokenType.Fn },
  { regex: /^\breturn\b/, tokenType: TokenType.Return },
  { regex: /^\bif\b/, tokenType: TokenType.If },
  { regex: /^\belse\b/, tokenType: TokenType.Else },
  { regex: /^\bwhile\b/, tokenType: TokenType.While },

  { regex: /^\bnumber\b/, tokenType: TokenType.NumberType },
  { regex: /^\bstring\b/, tokenType: TokenType.StringType },
  { regex: /^\bbool\b/, tokenType: TokenType.BooleanType },
  { regex: /^\bdynamic\b/, tokenType: TokenType.DynamicType },

  { regex: /^[a-zA-Z_][a-zA-Z0-9_]*/, tokenType: TokenType.Identifier },

  { regex: /^\/\/.*/, tokenType: TokenType.SingleLineComment },
  { regex: /^\/\/\/.*/, tokenType: TokenType.DocComment },

  // these are making problem with line numbers, because they are read as single line, no matter how many lines they span on....
  // { regex: /^\/\*[\s\S]*?\*\//, tokenType: TokenType.MultiLineComment },

  // operators
  { regex: /^[+-]/, tokenType: TokenType.AdditiveOperator },
  { regex: /^[*/%]/, tokenType: TokenType.MulitipicativeOperator },
  { regex: /^[><]=?/, tokenType: TokenType.RelationalOperator },
  { regex: /^[=!]=/, tokenType: TokenType.EqualityOperator },

  { regex: /^&&/, tokenType: TokenType.LogicGate },
  { regex: /^\|\|/, tokenType: TokenType.LogicGate },
  { regex: /^\band\b/, tokenType: TokenType.LogicGate },
  { regex: /^\bor\b/, tokenType: TokenType.LogicGate },

  { regex: /^:/, tokenType: TokenType.Colon },
  { regex: /^,/, tokenType: TokenType.Comma },
  { regex: /^\./, tokenType: TokenType.Dot },
  { regex: /^=/, tokenType: TokenType.Equals },
  { regex: /^!/, tokenType: TokenType.Exclamation },
  { regex: /^\(/, tokenType: TokenType.OpenParen },
  { regex: /^\)/, tokenType: TokenType.CloseParen },
  { regex: /^{/, tokenType: TokenType.OpenBrace },
  { regex: /^}/, tokenType: TokenType.CloseBrace },
  { regex: /^\[/, tokenType: TokenType.OpenBracket },
  { regex: /^\]/, tokenType: TokenType.CloseBracket },

  { regex: RegExp(`^${EOL}`), tokenType: TokenType.EOL },

  { regex: /^[^\S\r\n]+/, tokenType: TokenType.WhiteSpace }, // match one or more spaces and tabs
]
