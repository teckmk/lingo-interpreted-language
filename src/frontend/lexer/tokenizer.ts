/* eslint-disable @typescript-eslint/no-non-null-assertion */
import { InvalidIndentationError } from "../../error"
import { Spec, TokenType } from "./specs"

export type Token = {
  type: TokenType
  value: any
  column: number
  line: number
  code?: string
}

export default class Tokenizer {
  private MAX_QUEUE_LEN = 3

  private _spec: Spec[]
  private _tokens: Token[] = []
  private _tokenTypesQueue: TokenType[] = []

  private _cursor = 0
  private _line = 1 // current line
  private _tokenNumber = 1 // current column

  private _code = ""
  private _filename = ""

  private _tabSize = 0
  private _indentation = 0
  private _isIndentMode = false

  constructor(spec: Spec[], filename: string) {
    this._spec = spec
    this._filename = filename
  }

  tokenize(code: string) {
    this._code = code
    this._cursor = 0

    while (!this._isEOF && this._hasNext) this._readToken()

    this._pushToken({
      type: TokenType.EOF,
      value: "EOF",
      column: this._tokenNumber,
      line: this._line,
    })

    return this._tokens
  }

  private get _isEOF(): boolean {
    if (!this._code) return true

    return this._cursor == this._code.length
  }

  private get _hasNext(): boolean {
    if (!this._code || this._code.length == 0) return false

    return this._cursor < this._code.length
  }

  private _readToken() {
    const code = this._code.slice(this._cursor)

    for (const { regex, tokenType } of this._spec) {
      const tokenValue = this._matched(regex, code)

      // skip iteration if there's no match
      if (tokenValue == null) continue

      if (tokenType == TokenType.EOL) {
        this._line++
        this._tokenNumber = 0
      }

      if (tokenType == TokenType.WhiteSpace) {
        const indent = this._handleIndentation(tokenValue)
        if (indent) return
      }

      if (this._isIndentMode) {
        const dedent = this._handleDedentation(tokenValue, tokenType)
        if (dedent) return
      }

      if (
        tokenType == TokenType.EOL ||
        tokenType == TokenType.SingleLineComment ||
        tokenType == TokenType.WhiteSpace
      ) {
        this._updateQueue(tokenType)
        return
      }

      return this._pushToken({
        type: tokenType,
        value: tokenValue,
        column: this._tokenNumber,
        line: this._line,
        // code,
      })
    }

    throw new Error(
      `Unexpected token '${code[0]}' at ${this._line}:${this._tokenNumber} in ${this._filename}`
    )
  }

  private _matched(regex: RegExp, string: string) {
    const matched = regex.exec(string)

    if (matched == null) return null

    this._cursor += matched[0].length
    this._tokenNumber++

    return matched[0]
  }

  private _updateQueue(tokenType: TokenType) {
    const queueLen = this._tokenTypesQueue.length

    if (queueLen == this.MAX_QUEUE_LEN) this._tokenTypesQueue.shift()

    this._tokenTypesQueue.push(tokenType)
  }

  private _pushToken(token: Token) {
    this._tokens.push(token)
    this._updateQueue(token.type)
  }

  private _handleIndentation(tokenValue: string) {
    const q = this._tokenTypesQueue
    const whitespace = tokenValue.length

    if (q.length != this.MAX_QUEUE_LEN) return false

    // handle indent start
    if (q[1] != TokenType.Colon && q[2] != TokenType.EOL) return false

    // remove ':' token so it doesn't bother us in parser
    if (this._tokens.at(-1)?.type == TokenType.Colon) this._tokens.pop()

    if (this._tabSize == 0) this._tabSize = whitespace

    if (whitespace % this._tabSize != 0) throw InvalidIndentationError(this._tabSize, this._line)

    if (whitespace > this._indentation) {
      this._indentation += this._tabSize
      this._isIndentMode = true

      this._pushToken({
        type: TokenType.Indent,
        value: whitespace,
        column: this._tokenNumber,
        line: this._line,
      })

      return true
    }

    return false
  }

  private _handleDedentation(tokenValue: any, tokenType: TokenType) {
    if (this._tokenTypesQueue[2] != TokenType.EOL) return false

    if (tokenType == TokenType.WhiteSpace) {
      const whitespace = tokenValue.length

      if (whitespace < this._indentation) {
        this._indentation -= this._tabSize

        this._pushToken({
          type: TokenType.Dedent, // TokenType.Dedent
          value: tokenValue,
          column: this._tokenNumber,
          line: this._line,
        })

        return true
      }
    }

    if (this._indentation == this._tabSize) {
      this._indentation = 0
      this._isIndentMode = false

      this._pushToken({
        type: TokenType.Dedent, // TokenType.Dedent
        value: tokenValue,
        column: this._tokenNumber,
        line: this._line,
      })

      return true
    }

    return false
  }
}
