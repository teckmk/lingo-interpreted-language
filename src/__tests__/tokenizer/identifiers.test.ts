import { specs } from "../../frontend/lexer/specs";
import { Tokenizer } from "../../frontend/lexer/tokenizer";

describe("Tokenizer - Identifiers", () => {
  it("should tokenize identifiers", () => {
    const code = "x";
    const tokens = new Tokenizer(specs, "test").tokenize(code);
    expect(tokens).toMatchSnapshot();
  });

  it("should tokenize identifiers with capital letters", () => {
    const code = "X";
    const tokens = new Tokenizer(specs, "test").tokenize(code);
    expect(tokens).toMatchSnapshot();
  });

  it("should tokenize identifiers with mixed case", () => {
    const code = "xY";
    const tokens = new Tokenizer(specs, "test").tokenize(code);
    expect(tokens).toMatchSnapshot();
  });

  it("should tokenize identifiers with numbers", () => {
    const code = "x1";
    const tokens = new Tokenizer(specs, "test").tokenize(code);
    expect(tokens).toMatchSnapshot();
  });

  it("should tokenize identifiers with underscores", () => {
    const code = "x_";
    const tokens = new Tokenizer(specs, "test").tokenize(code);
    expect(tokens).toMatchSnapshot();
  });

  it("should tokenize identifiers with underscores and numbers", () => {
    const code = "x_1";
    const tokens = new Tokenizer(specs, "test").tokenize(code);
    expect(tokens).toMatchSnapshot();
  });

  it("should tokenize identifiers with numbers and trailing underscores", () => {
    const code = "x1_";
    const tokens = new Tokenizer(specs, "test").tokenize(code);
    expect(tokens).toMatchSnapshot();
  });

  it("should tokenize identifiers with leading underscores", () => {
    const code = "_x";
    const tokens = new Tokenizer(specs, "test").tokenize(code);
    expect(tokens).toMatchSnapshot();
  });
});
