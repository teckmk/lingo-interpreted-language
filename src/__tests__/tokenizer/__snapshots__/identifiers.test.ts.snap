// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tokenizer - Identifiers should tokenize identifiers 1`] = `
[
  {
    "position": {
      "end": {
        "column": 2,
        "line": 1,
      },
      "start": {
        "column": 1,
        "line": 1,
      },
    },
    "type": "IDENTIFIER",
    "value": "x",
  },
  {
    "position": {
      "end": {
        "column": -1,
        "line": 1,
      },
      "start": {
        "column": -1,
        "line": 1,
      },
    },
    "type": "END_OF_FILE",
    "value": "EOF",
  },
]
`;

exports[`Tokenizer - Identifiers should tokenize identifiers with capital letters 1`] = `
[
  {
    "position": {
      "end": {
        "column": 2,
        "line": 1,
      },
      "start": {
        "column": 1,
        "line": 1,
      },
    },
    "type": "TYPE_IDENTIFIER",
    "value": "X",
  },
  {
    "position": {
      "end": {
        "column": -1,
        "line": 1,
      },
      "start": {
        "column": -1,
        "line": 1,
      },
    },
    "type": "END_OF_FILE",
    "value": "EOF",
  },
]
`;

exports[`Tokenizer - Identifiers should tokenize identifiers with leading underscores 1`] = `
[
  {
    "position": {
      "end": {
        "column": 3,
        "line": 1,
      },
      "start": {
        "column": 1,
        "line": 1,
      },
    },
    "type": "IDENTIFIER",
    "value": "_x",
  },
  {
    "position": {
      "end": {
        "column": -1,
        "line": 1,
      },
      "start": {
        "column": -1,
        "line": 1,
      },
    },
    "type": "END_OF_FILE",
    "value": "EOF",
  },
]
`;

exports[`Tokenizer - Identifiers should tokenize identifiers with mixed case 1`] = `
[
  {
    "position": {
      "end": {
        "column": 3,
        "line": 1,
      },
      "start": {
        "column": 1,
        "line": 1,
      },
    },
    "type": "IDENTIFIER",
    "value": "xY",
  },
  {
    "position": {
      "end": {
        "column": -1,
        "line": 1,
      },
      "start": {
        "column": -1,
        "line": 1,
      },
    },
    "type": "END_OF_FILE",
    "value": "EOF",
  },
]
`;

exports[`Tokenizer - Identifiers should tokenize identifiers with numbers 1`] = `
[
  {
    "position": {
      "end": {
        "column": 3,
        "line": 1,
      },
      "start": {
        "column": 1,
        "line": 1,
      },
    },
    "type": "IDENTIFIER",
    "value": "x1",
  },
  {
    "position": {
      "end": {
        "column": -1,
        "line": 1,
      },
      "start": {
        "column": -1,
        "line": 1,
      },
    },
    "type": "END_OF_FILE",
    "value": "EOF",
  },
]
`;

exports[`Tokenizer - Identifiers should tokenize identifiers with numbers and trailing underscores 1`] = `
[
  {
    "position": {
      "end": {
        "column": 4,
        "line": 1,
      },
      "start": {
        "column": 1,
        "line": 1,
      },
    },
    "type": "IDENTIFIER",
    "value": "x1_",
  },
  {
    "position": {
      "end": {
        "column": -1,
        "line": 1,
      },
      "start": {
        "column": -1,
        "line": 1,
      },
    },
    "type": "END_OF_FILE",
    "value": "EOF",
  },
]
`;

exports[`Tokenizer - Identifiers should tokenize identifiers with underscores 1`] = `
[
  {
    "position": {
      "end": {
        "column": 3,
        "line": 1,
      },
      "start": {
        "column": 1,
        "line": 1,
      },
    },
    "type": "IDENTIFIER",
    "value": "x_",
  },
  {
    "position": {
      "end": {
        "column": -1,
        "line": 1,
      },
      "start": {
        "column": -1,
        "line": 1,
      },
    },
    "type": "END_OF_FILE",
    "value": "EOF",
  },
]
`;

exports[`Tokenizer - Identifiers should tokenize identifiers with underscores and numbers 1`] = `
[
  {
    "position": {
      "end": {
        "column": 4,
        "line": 1,
      },
      "start": {
        "column": 1,
        "line": 1,
      },
    },
    "type": "IDENTIFIER",
    "value": "x_1",
  },
  {
    "position": {
      "end": {
        "column": -1,
        "line": 1,
      },
      "start": {
        "column": -1,
        "line": 1,
      },
    },
    "type": "END_OF_FILE",
    "value": "EOF",
  },
]
`;
